File: frontend/src/types/audit.ts
export type AuditActorType = "system" | "user" | "atlas" | "integration" | string;

export type AuditEvent = {
  public_id: string;
  actor_type: AuditActorType;
  actor_user_public_id: string | null;
  actor_user_name: string | null;
  actor_role: string | null;
  entity_type: string;
  entity_public_id: string;
  action: string;
  before_data: Record<string, unknown> | null;
  after_data: Record<string, unknown> | null;
  metadata: Record<string, unknown>;
  created_at: string;
};

export type AuditActorOption = {
  public_id: string;
  name: string;
  role: string;
};

export type AuditStats = {
  total_events: number;
  user_events: number;
  system_events: number;
  atlas_events: number;
  integration_events: number;
};

export type AuditEventPage = {
  items: AuditEvent[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  stats: AuditStats;
  available_actor_types: string[];
  available_entity_types: string[];
  available_actions: string[];
  available_actors: AuditActorOption[];
};

File: app/schemas/audit.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class AuditEventRead(BaseModel):
    public_id: str
    actor_type: str
    actor_user_public_id: str | None
    actor_user_name: str | None = None
    actor_role: str | None = None
    entity_type: str
    entity_public_id: str
    action: str
    before_data: dict[str, object] | None
    after_data: dict[str, object] | None
    metadata: dict[str, object]
    created_at: datetime


class AuditActorOptionRead(BaseModel):
    public_id: str
    name: str
    role: str


class AuditStatsRead(BaseModel):
    total_events: int
    user_events: int
    system_events: int
    atlas_events: int
    integration_events: int


class AuditEventPageRead(BaseModel):
    items: list[AuditEventRead]
    total: int
    page: int
    page_size: int
    total_pages: int
    stats: AuditStatsRead
    available_actor_types: list[str]
    available_entity_types: list[str]
    available_actions: list[str]
    available_actors: list[AuditActorOptionRead]
